invenio/modules/deposit/cache.py

```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015 CERN.
+#
+# Invenio is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Implementation of allowed deposit type caching."""
+
+from invenio.legacy.miscutil.data_cacher import DataCacher, DataCacherProxy
+
+
+class DepositTypeDataCacher(DataCacher):
+    def __init__(self):
+        def cache_filler():
+            from invenio.modules.access.control import acc_get_action_id
+            from invenio.modules.access.models import AccAuthorization, \
+                AccARGUMENT
+
+            action_id = acc_get_action_id('usedeposit')
+
+            return [auth[0] for auth in AccAuthorization.query.join(
+                AccAuthorization.argument
+            ).filter(
+                AccARGUMENT.keyword == 'type',
+                AccAuthorization.id_accACTION == action_id
+            ).values(AccARGUMENT.value)]
+
+        def timestamp_verifier():
+            from invenio.legacy.dbquery import get_table_update_time
+            return max(get_table_update_time('accROLE_accACTION_accARGUMENT'),
+                       get_table_update_time('accARGUMENT'))
+
+        DataCacher.__init__(self, cache_filler, timestamp_verifier)
+
+
+deposit_type_cacher = DataCacherProxy(DepositTypeDataCacher)
+
+
+def get_authorized_deposition_types(user_info, recreate_cache_if_needed=True):
+    """Return a list of allowed deposition types for a certain user."""
+    from invenio.modules.access.engine import acc_authorize_action
+
+    if recreate_cache_if_needed:
+        deposit_type_cacher.recreate_cache_if_needed()
+    ret = []
+
+    auths = acc_authorize_action(
+        user_info,
+        'usedeposit',
+        batch_args=True,
+        collection=deposit_type_cacher.cache
+    )
+
+    for collection, auth in zip(deposit_type_cacher.cache, auths):
+        if auth[0] == 0:
+            ret.append(collection)
+    return ret
```

Comment:
Can you keep the new line character here(`\n`) (not new empty line `\n\n` or none `\0`). Thanks